p-ata/src/processor.rs

```diff
@@ -0,0 +1,974 @@
+/// * `Some((address, bump))` - A better bump was found, returns the canonical address and bump
+///
+/// ## Security
+///
+/// This function is critical for preventing PDA canonicality attacks. It ensures
+/// that only the highest valid bump can be used, maintaining deterministic
+/// address derivation across all clients.
+#[inline(always)]
+pub(crate) fn reject_if_better_valid_bump_exists(
+    seeds: &[&[u8]; 3],
+    program_id: &Pubkey,
+    expected_bump: u8,
+) -> Result<(), ProgramError> {
+    // Optimization: Only verify no better bump exists. Don't require expected_bump to
+    // yield an off-curve address. This saves significant compute units while still
+    // preventing non-canonical addresses.
+    //
+    // Caller must ensure that `expected_bump` is off-curve, either via downstream failure
+    // (i.e. syscalls that will fail) or by calling `is_off_curve`.
+    let mut better_bump = 255;
+    while better_bump > expected_bump {
+        let maybe_better_address = derive_address::<3>(seeds, Some(better_bump), program_id);
+        if is_off_curve(&maybe_better_address) {
+            log!("Canonical address does not match provided address. Canonical bump is {}, with address {}.", better_bump, &maybe_better_address);
+            return Err(ProgramError::InvalidInstructionData);
+        }
+        better_bump -= 1;
+    }
+    Ok(())
+}
+
+/// Accounts:
+/// [0] payer
+/// [1] associated_token_account_to_create
+/// [2] wallet
+/// [3] mint
+/// [4] system_program
+/// [5] token_program
+/// [6] rent_sysvar
+///
+/// For Token-2022 accounts, create the account with the correct size
+/// and call InitializeImmutableOwner followed by InitializeAccount3.
+#[inline(always)]
+pub(crate) fn process_create_associated_token_account(
+    program_id: &Pubkey,
+    create_accounts: &CreateAccounts,
+    expected_bump: Option<u8>,
+    known_token_account_len: Option<usize>,
+) -> ProgramResult {
+    let bump = match expected_bump {
+        Some(provided_bump) => {
+            // Check if a better canonical bump exists
+            reject_if_better_valid_bump_exists(
+                &[
+                    create_accounts.wallet.key().as_ref(),
+                    create_accounts.token_program.key().as_ref(),
+                    create_accounts.mint.key().as_ref(),
+                ],
+                program_id,
+                provided_bump,
+            )?;
+            provided_bump
+        }
+        None => {
+            let (_address, computed_bump) = derive_canonical_ata_pda(
+                create_accounts.wallet.key(),
+                create_accounts.token_program.key(),
+                create_accounts.mint.key(),
+                program_id,
+            );
+            computed_bump
+        }
+    };
+
+    let space = resolve_token_account_space(
+        create_accounts.token_program,
+        create_accounts.mint,
+        known_token_account_len,
+    )?;
+
+    match create_accounts.rent_sysvar {
```

Comment:
The `rent_sysvar` seems to be used only to compute the minimal rent but at this point you already have the `space`. You could calculate the rent requirement here and replace the `match` statement with something like:
```rust
let minimum_balance = if let Some(rent_sysvar_info) = create_accounts.rent_sysvar {
        let rent = unsafe { Rent::from_account_info_unchecked(rent_sysvar_info)? };
        rent.minimum_balance(space)
} else {
        Rent::get()?.minimum_balance(space)
};
create_and_initialize_ata(...)
```